src/main/java/org/littleshoot/proxy/extras/SelfSignedSslEngineSource.java

```diff
@@ -70,6 +98,83 @@ public SelfSignedSslEngineSource() {
     public SSLEngine newSslEngine() {
         return sslContext.createSSLEngine();
     }
+    
+    public SSLEngine clientSslEngineFor(SSLSession remoteServerSslSession) {
+    	try{
+
+    		String sslContextKey = remoteServerSslSession.getPeerHost() +":" + remoteServerSslSession.getPeerPort();
+    		SSLContext newSslContext = null;
+    		if(!sslContextCache.contains(sslContextKey) || (newSslContext = sslContextCache.get(sslContextKey).get()) == null ){
+    			
+    			X509Certificate[] remoteServerCertChain = remoteServerSslSession.getPeerCertificateChain();
+        		X509CertImpl remoteServerCert = new X509CertImpl(remoteServerCertChain[0].getEncoded());
+        		X509CertInfo remoteServerCertInfo = (X509CertInfo) remoteServerCert.get(X509CertImpl.NAME + "." + X509CertImpl.INFO);
+
+        		newSslContext = SSLContext.getInstance("SSL");
+        		// Get our key pair and our own DN (not the remote server's DN) from the keystore.
+        		X509CertImpl caCert = new X509CertImpl(keyStore.getCertificate(ALIAS).getEncoded());
+        		PrivateKey caPrivateKey = (PrivateKey) keyStore.getKey(ALIAS, PASSWORD.toCharArray());
+        		PublicKey caPublicKey = caCert.getPublicKey();
+
+        		X509CertInfo caCertInfo = (X509CertInfo) caCert.get(X509CertImpl.NAME + "." + X509CertImpl.INFO);
+        		X500Name issuer = (X500Name) caCertInfo.get(X509CertInfo.SUBJECT + "." + CertificateIssuerName.DN_NAME);
+
+        		//modify certificats
+        		Random random = new Random();
+        		remoteServerCertInfo.set(X509CertInfo.VERSION, new CertificateVersion(CertificateVersion.V3));
+        		remoteServerCertInfo.set(X509CertInfo.ISSUER + "." + CertificateSubjectName.DN_NAME, issuer);
+        		remoteServerCertInfo.set(X509CertInfo.KEY, new CertificateX509Key(caPublicKey));
+        		remoteServerCertInfo.set(X509CertInfo.SERIAL_NUMBER, new CertificateSerialNumber(new BigInteger(random.nextLong()+"")));
```

Comment:
In theory this can lead to serial number collisions.  That's likely not an issue, but per the X509 spec different certificates from the same CA are supposed to have different serial numbers.